src/modules/mc_pos_control/mc_pos_control_main.cpp

```diff
@@ -1292,784 +1308,797 @@ void MulticopterPositionControl::control_auto(float dt)
-	_att_sp.landing_gear = -1.0f;
+		if (_params.alt_mode == 1 && _local_pos.dist_bottom_valid) {
+			_vel(2) = -_local_pos.dist_bottom_rate;
 
+		} else {
+			_vel(2) = _local_pos.vz;
+		}
+	}
 
-	matrix::Dcmf R;
-	R.identity();
+	_vel_err_d(0) = _vel_x_deriv.update(-_vel(0));
+	_vel_err_d(1) = _vel_y_deriv.update(-_vel(1));
+	_vel_err_d(2) = _vel_z_deriv.update(-_vel(2));
+}
 
-	/* wakeup source */
-	px4_pollfd_struct_t fds[1];
+void
+MulticopterPositionControl::do_position_control(float dt)
+{
 
-	fds[0].fd = _local_pos_sub;
-	fds[0].events = POLLIN;
+	_vel_ff.zero();
 
-	while (!_task_should_exit) {
-		/* wait for up to 20ms for data */
-		int pret = px4_poll(&fds[0], (sizeof(fds) / sizeof(fds[0])), 20);
+	/* by default, run position/altitude controller. the control_* functions
+	 * can disable this and run velocity controllers directly in this cycle */
+	_run_pos_control = true;
+	_run_alt_control = true;
 
-		/* timed out - periodic check for _task_should_exit */
-		if (pret == 0) {
-			// Go through the loop anyway to copy manual input at 50 Hz.
-		}
+	/* select control source */
+	if (_control_mode.flag_control_manual_enabled) {
+		/* manual control */
+		control_manual(dt);
+		_mode_auto = false;
 
-		/* this is undesirable but not much we can do */
-		if (pret < 0) {
-			warn("poll error %d, %d", pret, errno);
-			continue;
+	} else if (_control_mode.flag_control_offboard_enabled) {
+		/* offboard control */
+		control_offboard(dt);
+		_mode_auto = false;
+
+	} else {
+		/* AUTO */
+		control_auto(dt);
+	}
+
+	/* weather-vane mode for vtol: disable yaw control */
+	if (!_control_mode.flag_control_manual_enabled && _pos_sp_triplet.current.disable_mc_yaw_control == true) {
+		_att_sp.disable_mc_yaw_control = true;
+
+	} else {
+		/* reset in case of setpoint updates */
+		_att_sp.disable_mc_yaw_control = false;
+	}
+
+	if (!_control_mode.flag_control_manual_enabled && _pos_sp_triplet.current.valid
+	    && _pos_sp_triplet.current.type == position_setpoint_s::SETPOINT_TYPE_IDLE) {
+		/* idle state, don't run controller and set zero thrust */
+		_R_setpoint.identity();
+
+
+		matrix::Quatf qd = _R_setpoint;
+		memcpy(&_att_sp.q_d[0], qd.data(), sizeof(_att_sp.q_d));
+		_att_sp.q_d_valid = true;
+
+		_att_sp.roll_body = 0.0f;
+		_att_sp.pitch_body = 0.0f;
+		_att_sp.yaw_body = _yaw;
+		_att_sp.thrust = 0.0f;
+
+		_att_sp.timestamp = hrt_absolute_time();
+
+		/* publish attitude setpoint */
+		if (_att_sp_pub != nullptr) {
+			orb_publish(_attitude_setpoint_id, _att_sp_pub, &_att_sp);
+
+		} else if (_attitude_setpoint_id) {
+			_att_sp_pub = orb_advertise(_attitude_setpoint_id, &_att_sp);
 		}
 
-		poll_subscriptions();
+	} else if (_control_mode.flag_control_manual_enabled
+		   && _vehicle_land_detected.landed) {
+		/* don't run controller when landed */
+		_reset_pos_sp = true;
+		_reset_alt_sp = true;
+		_do_reset_alt_pos_flag = true;
+		_mode_auto = false;
+		_reset_int_z = true;
```

Comment:
we could add a function that resets all the flags such that we dont have to rewrite what already was done in the initialization part